src/appleseed/renderer/kernel/lighting/volumelightingintegrator.cpp

```diff
@@ -0,0 +1,601 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "renderer/kernel/lighting/volumelightingintegrator.h"
+
+// appleseed.renderer headers.
+#include "renderer/kernel/lighting/backwardlightsampler.h"
+#include "renderer/kernel/lighting/directlightingintegrator.h"
+#include "renderer/kernel/lighting/tracer.h"
+#include "renderer/kernel/shading/shadingcomponents.h"
+#include "renderer/kernel/shading/shadingcontext.h"
+#include "renderer/kernel/shading/shadingpoint.h"
+#include "renderer/modeling/volume/volume.h"
+#include "renderer/modeling/edf/edf.h"
+#include "renderer/modeling/light/light.h"
+#include "renderer/modeling/material/material.h"
+#include "renderer/modeling/scene/visibilityflags.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/rr.h"
+#include "foundation/math/sampling/mappings.h"
+#include "foundation/math/scalar.h"
+
+// Standard headers.
+#include <cassert>
+#include <cmath>
+
+using namespace foundation;
+
+namespace renderer
+{
+
+namespace
+{
+    struct EquiangularSampler
+    {
+      public:
+        EquiangularSampler(
+            const LightSample&          light_sample,
+            const ShadingRay&           volume_ray,
+            const ShadingContext&       shading_context,
+            SamplingContext&            sampling_context)
+          : m_shading_context(shading_context)
+          , m_sampling_context(sampling_context)
+        {
+            Vector3d emission_position;
+            if (light_sample.m_light != nullptr)
+            {
+                m_sampling_context.split_in_place(2, 1);
+                Vector3d emission_direction;  // not used
+                Spectrum light_value(Spectrum::Illuminance);
+                float probability;
+                light_sample.m_light->sample(
+                    m_shading_context,
+                    light_sample.m_light_transform,
+                    m_sampling_context.next2<Vector2d>(),
+                    emission_position,
+                    emission_direction,
+                    light_value,
+                    probability);
+            }
+            else
+                emission_position = light_sample.m_point;
+
+            const Vector3d light_dir = emission_position - volume_ray.m_org;
+
+            m_origin_to_center = dot(light_dir, volume_ray.m_dir);
+
+            m_center_to_light =
+                std::sqrt(square_norm(light_dir) - square(m_origin_to_center));
```

Comment:
I checked the consistency between equiangular sampling and exponential sampling and even had the test scenes with both (before I merged `m_exponential_samples` and `m_equiangular_samples` together)

Reply:
I checked the consistency between equiangular sampling and exponential sampling and even had corresponding test scenes (before I merged `m_exponential_samples` and `m_equiangular_samples` together). So the math is certainly okay.
We still can check consistency between exponential-only sampling and combined, since we have this "Optimize for..." option.
Probably, I just have confusing names and lack of comments...